les/fetcher.go

```diff
@@ -17,871 +17,469 @@
-	requested         map[uint64]fetchRequest
-	deliverChn        chan fetchResponse
-	timeoutChn        chan uint64
-	requestTriggered  bool
-	requestTrigger    chan struct{}
-	lastTrustedHeader *types.Header
+	outOrderThreshold = 3         // The maximum limit the server is allowed to send out_of_order announces
+	timeoutThreshold  = 5         // The maximum limit the server is allowed for timeout requests
+	expirationRate    = time.Hour // The linear expiration rate of the _bad_ behaviors statistic
+)
 
-	closeCh chan struct{}
-	wg      sync.WaitGroup
+// announce represents an new block announcement from the les server.
+type announce struct {
+	data   *announceData
+	trust  bool
+	peerid enode.ID
 }
 
-// fetcherPeerInfo holds fetcher-specific information about each active peer
-type fetcherPeerInfo struct {
-	root, lastAnnounced *fetcherTreeNode
-	nodeCnt             int
-	confirmedTd         *big.Int
-	bestConfirmed       *fetcherTreeNode
-	nodeByHash          map[common.Hash]*fetcherTreeNode
-	firstUpdateStats    *updateStatsEntry
+// request represents a record when the header request is sent.
+type request struct {
+	reqid  uint64
+	peerid enode.ID
+	sendAt time.Time
 }
 
-// fetcherTreeNode is a node of a tree that holds information about blocks recently
-// announced and confirmed by a certain peer. Each new announce message from a peer
-// adds nodes to the tree, based on the previous announced head and the reorg depth.
-// There are three possible states for a tree node:
-// - announced: not downloaded (known) yet, but we know its head, number and td
-// - intermediate: not known, hash and td are empty, they are filled out when it becomes known
-// - known: both announced by this peer and downloaded (from any peer).
-// This structure makes it possible to always know which peer has a certain block,
-// which is necessary for selecting a suitable peer for ODR requests and also for
-// canonizing new heads. It also helps to always download the minimum necessary
-// amount of headers with a single request.
-type fetcherTreeNode struct {
-	hash             common.Hash
-	number           uint64
-	td               *big.Int
-	known, requested bool
-	parent           *fetcherTreeNode
-	children         []*fetcherTreeNode
+// response represents a response packet from network as well as a channel
+// to return all un-requested data.
+type response struct {
+	reqid   uint64
+	headers []*types.Header
+	peerid  enode.ID
+	remain  chan []*types.Header
 }
 
-// fetchRequest represents a header download request
-type fetchRequest struct {
-	hash    common.Hash
-	amount  uint64
-	peer    *serverPeer
-	sent    mclock.AbsTime
-	timeout bool
+// fetcherPeer holds the fetcher-specific information for each active peer
+type fetcherPeer struct {
+	latest   *announceData            // The latest announcement sent from the peer
+	timeout  utils.LinearExpiredValue // The counter of all timeout requests made
```

Comment:
Though this `LinearExpiredValue` is nice I am not sure we should use it here. "Hard" timeouts (10s timeout) are always dropped by the ODR retriever, that is practically a broken connection and we should not waste time trying to request from it.

Reply:
dropped